### PR TITLE
Remove summer heat demand and distributes excess demand over the profile

### PR DESCRIPTION
**Pull request by MaGering**
Fix #Issue https://github.com/greco-project/pvcompare/issues/136

With this PR heat demand during summer is removed from an implemented daily mean temperature (right now 15 °C) on. This excess heat demand is then distributed equally over the remaining heat demand.

The function is used as an enhancement of the [demandlib](https://github.com/oemof/demandlib), which calculates heat demands only above zero. For modelling the heat demand without warm water of dwelling and houses an approach is needed to approximate to the real heating behavior.

The following steps were realized, as well (if applies):
- [x] Use in-line comments to explain your code
- [x] Write docstrings to your code
- [x] For new functionalities: Explain in readthedocs
- [x] Write test(s) for your new patch of code
- :x: Update the CHANGELOG.md (let's start this after the first release)
- [x] Apply black (`black . --exclude docs/`)

*Please mark above checkboxes as following:*
- [ ] Open
- [x] Done

:x: Check not applicable to this PR

<sub>*For more information on how to contribute check the [CONTRIBUTING.md](https://github.com/rl-institut/mvs_eland/blob/dev/CONTRIBUTING.md).*<sub>

### CODE REVIEW

**Review by SabineHaas — CHANGES_REQUESTED** [review]
Thank you @MaGering for solving this issue! 

- Could you add a docstring to your new function? There's a good example e.g. in `shift_working_hours()`.
- Please add tests for `adjust_heat_demand()`. If you need examples you can look into the other tests functions in directory `tests`. Your tests should be located in `test_demand.py`.
- You could add the temperature that marks the heating period (in your case 15 °C) as a parameter again if you like - so that it's easier to select a different temperature.

**Comment by MaGering** [comment]
> Thank you @MaGering for solving this issue!
> 
> * Could you add a docstring to your new function? There's a good example e.g. in `shift_working_hours()`.
> * Please add tests for `adjust_heat_demand()`. If you need examples you can look into the other tests functions in directory `tests`. Your tests should be located in `test_demand.py`.
> * You could add the temperature that marks the heating period (in your case 15 °C) as a parameter again if you like - so that it's easier to select a different temperature.


- [ ] Add a docstring to new function
- [ ] Add tests for `adjust_heat_demand()` located in `test_demand.py`
- [ ] Add limit temperature for heating period as input parameter so that it's easier to select a different temperature

**Comment by MaGering** [comment]
>     * You could add the temperature that marks the heating period (in your case 15 °C) as a parameter again if you like - so that it's easier to select a different temperature.

So I would include it as an input parameter in building_parameters.csv?

**Comment by SabineHaas** [comment]
> > ```
> > * You could add the temperature that marks the heating period (in your case 15 °C) as a parameter again if you like - so that it's easier to select a different temperature.
> > ```
> 
> So I would include it as an input parameter in building_parameters.csv?

At first I thought about `heat_pumps_and_chillers.csv` but maybe `building_parameters.csv` fits better! I don't have a preference. ;)

**Comment by MaGering** [comment]
I extended the adjustment of the hourly heat demand in case warm water is included:

By setting the parameter `include_warm_water` to True, warm water is included in the calculations. In this case the hourly heat demand of space heating is separated from the hourly heat demand of warm water. The hourly heat demand of space heating **only** is then adjusted using `adjust_heat_demand(...)`. To finally obtain the total adjusted hourly heat demand the hourly heat demand of warm water **only** is added.

**Comment by MaGering** [comment]
Here you can see an example plot of the adjusted hourly heat demand with and without warm water:

Heat demand without warm water
Heat demand with warm water

File: pvcompare/demand.py [hunk 1]
```diff
@@ -305,39 +305,53 @@ def calculate_heat_demand(
...
 
     filename_population = bp.at["filename_country_population", "value"]
     filename1 = os.path.join(input_directory, filename_population)
     populations = pd.read_csv(filename1, index_col=0, sep=",")
+
     # convert Mtoe in kWh
+    # Heat demand of residential for space heating
     heat_demand = (
         total_SH.at[country, str(year)]  #       + total_WH.at[country, str(year)]
         - electr_SH.at[country, str(year)]  #       - electr_WH.at[country, str(year)]
```
Comment (SabineHaas) [inline pvcompare/demand.py:336]: You can delete the comments now

File: pvcompare/demand.py [hunk 2]
```diff
@@ -305,39 +305,53 @@ def calculate_heat_demand(
...
     ) * 11630000000
     annual_heat_demand_per_population = (
         heat_demand / populations.at[country, str(year)]
     ) * population
+    # Heat demand of households for water heating
+    heat_demand_ww = (
+        total_WH.at[country, str(year)]  # + total_WH.at[country, str(year)]
+        - electr_WH.at[country, str(year)]  # - electr_WH.at[country, str(year)]
```
Comment (SabineHaas) [inline pvcompare/demand.py:344]: also here (comments)

File: pvcompare/demand.py [hunk 3]
```diff
@@ -305,39 +305,53 @@ def calculate_heat_demand(
...
+    ) * 11630000000
+
```
Comment (SabineHaas) [inline pvcompare/demand.py:346]:
```suggestion
```

File: pvcompare/demand.py [hunk 4]
```diff
@@ -305,39 +305,53 @@ def calculate_heat_demand(
...
+    annual_heat_demand_ww_per_population = (
+        heat_demand_ww / populations.at[country, str(year)]
+    ) * population
 
     # Multi family house (mfh: Mehrfamilienhaus)
+    include_warm_water = False  # Set true to include warm water
```
Comment (SabineHaas) [inline pvcompare/demand.py:352]: Could you add `include_warm_water` as a parameter of this function? and/or it could be a parameter in an input csv so that the user can choose. Does this make sense?

File: pvcompare/demand.py [hunk 5]
```diff
@@ -375,6 +428,52 @@ def calculate_heat_demand(
     return shifted_heat_demand
 
 
+def adjust_heat_demand(temperature, heating_limit_temp, demand):
+    """
+    Adjust the hourly heat demand setting a limit to the temperature of heating
```
Comment (SabineHaas) [inline pvcompare/demand.py:433]:
```suggestion
    Adjust the hourly heat demand setting a limit to the temperature of heating
    
```

File: tests/test_demand.py [hunk 6]
```diff
@@ -123,6 +136,16 @@ def test_calculate_heat_demand(self):
...
 
+    def test_adjust_heat_demand(self):
+
+        result = adjust_heat_demand(
+            temperature=self.weather["temp_air"],
+            heating_limit_temp=self.heating_lim_temp,
+            demand=self.heat_load["kWh"],
+        )
+
+        assert result.sum() == 10969639.113628691
```
Comment (SabineHaas) [inline tests/test_demand.py:147]:
```suggestion
        assert result.sum() == heat_load["kWh"].sum()
```
I think here you test, that the sum stays the same (because the load from summer times is distributed to the rest of the year). Could you also write a test that checks that expected values are set to zero please?

**Review by SabineHaas — CHANGES_REQUESTED** [review]
This is a very handy feature!

Could you add a test for `calculate_heat_demand()` where warm water is included please (once there's the option to in- or exclude it)?

File: pvcompare/demand.py (hunk 1 above)
Comment (MaGering) [reply pvcompare/demand.py]: Fixed with https://github.com/greco-project/pvcompare/pull/137/commits/c09bbce08c335434419b702f71a31961257f58a4

File: pvcompare/demand.py (hunk 2 above)
Comment (MaGering) [reply pvcompare/demand.py]: Fixed with https://github.com/greco-project/pvcompare/pull/137/commits/c09bbce08c335434419b702f71a31961257f58a4

File: pvcompare/demand.py (hunk 4 above)
Comment (MaGering) [reply pvcompare/demand.py]: Yes, it does! :)
Adapted with https://github.com/greco-project/pvcompare/pull/137/commits/06b0cb0fbc11ae55bd433157fab97d95d615c2b5

**Comment by MaGering** [comment]
> Could you add a test for `calculate_heat_demand()` where warm water is included please (once there's the option to in- or exclude it)?

Done with commit https://github.com/greco-project/pvcompare/pull/137/commits/83284c8e51c4c3f4163b7e9dfb8728f469e68e61

File: tests/test_demand.py (hunk 6 above)
Comment (MaGering) [reply tests/test_demand.py]: Good idea! :)

Done with commit https://github.com/greco-project/pvcompare/pull/137/commits/289894fd4fa37b858f95664c1c21b6d0bda90064

**Review by SabineHaas — APPROVED** [review]

**Review by c-moeller — APPROVED** [review]

**Review by Piranias — APPROVED** [review]
